### DISCUSSION TITLE
getServerSideProps does not work after "npm run build + npm run start"

### DISCUSSION THREAD

**Discussion by rabieeali** [vercel/next.js#55312] [App Router]
why does this code (getServerSideProps) work in development mode but it DOES NOT work in production? :
I am not getting redirected to "/auth/login/?redirectPath=/checkout" when there is no cookie named "token" in production build
this is nextJS version 13.2.4 by the way

```import { GetServerSidePropsContext, NextPage } from 'next'
import CartPagesLayout from '@/components/layout/Cart'
import { useCartQuery } from '@/hooks'
import { Cart } from '@/types/interface/shopping-cart'
import Link from 'next/link'
import cookie from 'cookie'

const CheckoutPage: NextPage = () => {
  const { data } = useCartQuery()
  const currentCart: Cart = data?.data

  return (
    <CartPagesLayout>
      {currentCart === null && (
        <div className="flex flex-col justify-center items-center  my-10 xl:my-20 text-base gap-4">
          <p className=" text-gray">
            no item in cart!
          </p>
          <Link href="/" className="px-4 py-2 rounded-full bg-yellow text-white">
            back to homepage
          </Link>
        </div>
      )}
      <h1>this is Checkout Page</h1>
    </CartPagesLayout>
  )
}

export default CheckoutPage

export async function getServerSideProps(ctx: GetServerSidePropsContext) {
  if (ctx?.req?.headers?.cookie) {
    const { token } = cookie.parse(ctx?.req?.headers?.cookie)
    if (token) {
      return {
        props: {},
      }
    } else {
      return {
        props: {},
        redirect: {
          destination: '/auth/login/?redirectPath=/checkout',
          permanent: false,
        },
      }
    }
  } else {
    return {
      props: {},
      redirect: {
        destination: '/auth/login/?redirectPath=/checkout',
        permanent: false,
      },
    }
  }
}
```

**Comment by icyJoseph** [comment]
Have you been able to debug this? 

Does GSSP not run at all? or are the cookies given issues? How are the cookies set? Is the cookie path correct?

**Reply by rabieeali** [reply]
> Have you been able to debug this? 
> 
> Does GSSP not run at all? or are the cookies given issues? How are the cookies set? Is the cookie path correct? 

everything's correct, I even checked it with ChatGPT but it doesn't work on after build!

there is also another bug I ran into!

sometimes when I get back from a route that has getServerSideProps fetched data with Link component , I get redirected to my error boundary page!

it's weird! isn't the Link component supposed to show the page with cache data?! why do I always fetch the data when navigating with Link?

**Reply by icyJoseph** [reply]
Are you perhaps using PWA? `next-pwa`? 

This is using the Pages Router right? Not the App Router?

**Reply by rabieeali** [reply]
> Are you perhaps using PWA? `next-pwa`? 
> 
> This is using the Pages Router right? Not the App Router?

yes sir I am using next-pwa package version 5.6.0 and here is my next.config.js code:

```
/* @type {import('next').NextConfig} /
const runtimeCaching = require('next-pwa/cache')

const withPWA = require('next-pwa')({
  dest: 'public',
  register: true,
  skipWaiting: true,
  runtimeCaching,
  buildExcludes: [/middleware-manifest.json$/],
  disableDevLogs: true,
  mode: process.env.NODE_ENV === 'development' ? 'development' : 'production',
  disable: process.env.NODE_ENV === 'development',
})

const nextConfig = withPWA({
  reactStrictMode: true,

  images: {
    domains: ['my-first-domain', my-second-domain'],
  },
  webpack(config) {
    config.module.rules.push({
      test: /\.svg$/i,
      issuer: /\.[jt]sx?$/,
      use: ['@svgr/webpack'],
    })

    return config
  },
})

module.exports = nextConfig

```

**Reply by icyJoseph** [reply]
Hi, 

That's the problem. A PWA with GSSP, requires special care.

Please refer to these answers: 

- https://github.com/vercel/next.js/discussions/52024#discussioncomment-6325542
- https://github.com/vercel/next.js/discussions/51454#discussioncomment-6218798
- https://github.com/vercel/next.js/discussions/42405#discussioncomment-4046461
- https://github.com/vercel/next.js/discussions/42397#discussioncomment-4045174

## Visibility repost

Summary for visibility

next-pwa, by default has something called stale-while-revalidate, https://developer.chrome.com/docs/workbox/modules/workbox-strategies/#stale-while-revalidate, which for GSSP, is a problem.

You need to use https://developer.chrome.com/docs/workbox/modules/workbox-strategies/#network-first-network-falling-back-to-cache:

```js
/** @type {import('next').NextConfig} */
const runtimeCaching = require("next-pwa/cache");

const nextDataIndex = runtimeCaching.findIndex(
  (entry) => entry.options.cacheName === "next-data"
);

if (nextDataIndex !== -1) {
  runtimeCaching[nextDataIndex].handler = "NetworkFirst";
} else {
  throw new Error("Failed to find next-data object in runtime caching");
}

const withPWA = require("next-pwa")({
  dest: "public",
  runtimeCaching,
  register: true,
  skipWaiting: true,
  buildExcludes: [/middleware-manifest.json$/],
  disable: process.env.NODE_ENV === 'development'
})

module.exports = withPWA({
  reactStrictMode: false
});
```
